File: system_app_front_end/lib/areas/objects/data/image_payload.dart
/// Image object payload: one picture, or a row of panes.
class ImageObjectPayload {
  ImageObjectPayload._();

  static List<Map<String, String>> panesOf(Map<String, dynamic>? payload) {
    final raw = payload ?? const <String, dynamic>{};
    final images = raw['images'];
    if (images is List && images.isNotEmpty) {
      final panes = <Map<String, String>>[];
      for (final item in images) {
        if (item is! Map) continue;
        panes.add({
          'url': '${item['url'] ?? item['path'] ?? ''}'.trim(),
          'caption': '${item['caption'] ?? ''}',
        });
      }
      if (panes.isNotEmpty) return panes;
    }
    return [
      {
        'url': '${raw['url'] ?? raw['path'] ?? ''}'.trim(),
        'caption': '${raw['caption'] ?? ''}',
      },
    ];
  }

  /// First pane is also mirrored on [url] / [caption] for older readers.
  static Map<String, dynamic> mirrored(
    List<Map<String, String>> panes, {
    Map<String, dynamic>? existing,
  }) {
    final cleaned = [
      for (final p in panes)
        {
          'url': (p['url'] ?? '').trim(),
          'caption': p['caption'] ?? '',
        },
    ];
    if (cleaned.isEmpty) {
      cleaned.add({'url': '', 'caption': ''});
    }
    final first = cleaned.first;
    final out = <String, dynamic>{
      ...?existing,
      'url': first['url'],
      'caption': first['caption'],
    };
    if (cleaned.length > 1) {
      out['images'] = cleaned;
    } else {
      out.remove('images');
    }
    return out;
  }

  static Map<String, dynamic> merge(
    Map<String, dynamic>? keeper,
    Map<String, dynamic>? absorbed,
  ) {
    return mirrored(
      [...panesOf(keeper), ...panesOf(absorbed)],
      existing: keeper,
    );
  }
}

File: system_app_back_end/areas/objects/services/image_payload.py
"""Image object payload: one picture, or a row of panes."""

from __future__ import annotations

from typing import Any


def panes_of(payload: dict[str, Any] | None) -> list[dict[str, Any]]:
    raw = dict(payload or {})
    images = raw.get("images")
    if isinstance(images, list) and images:
        panes: list[dict[str, Any]] = []
        for item in images:
            if not isinstance(item, dict):
                continue
            url = str(item.get("url") or item.get("path") or "").strip()
            caption = str(item.get("caption") or "")
            panes.append({"url": url, "caption": caption})
        if panes:
            return panes
    return [
        {
            "url": str(raw.get("url") or raw.get("path") or "").strip(),
            "caption": str(raw.get("caption") or ""),
        }
    ]


def mirrored(
    panes: list[dict[str, Any]],
    *,
    width: Any = None,
    look: str | None = None,
) -> dict[str, Any]:
    """First pane on url/caption so older readers still see one picture."""
    cleaned = [
        {
            "url": str(p.get("url") or "").strip(),
            "caption": str(p.get("caption") or ""),
        }
        for p in panes
        if isinstance(p, dict)
    ]
    if not cleaned:
        cleaned = [{"url": "", "caption": ""}]
    first = cleaned[0]
    out: dict[str, Any] = {
        "url": first["url"],
        "caption": first["caption"],
    }
    if width is not None and width != "":
        out["width"] = width
    if look:
        out["look"] = look
    if len(cleaned) > 1:
        out["images"] = cleaned
    return out
